/* -------------------------------------------------------------------------- */
/* Copyright 2002-2025, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */
syntax = "proto3";

import "shared.proto";

package one.tmpl;

option go_package = "github.com/OpenNebula/one/src/oca/go/src/goca/api/template";

message AllocateRequest
{
  string session_id = 1;
  string template   = 2;
}

message DeleteRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool recursive    = 3;
}

message InfoRequest
{
  string session_id = 1;
  int32 oid         = 2;
  bool extended     = 3;
  bool decrypt      = 4;
}

message UpdateRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string template   = 3;
  int32 append      = 4;
}

message RenameRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string name       = 3;
}

message ChmodRequest
{
  string session_id   = 1;
  int32 oid           = 2;
  sint32 user_use     = 3;
  sint32 user_manage  = 4;
  sint32 user_admin   = 5;
  sint32 group_use    = 6;
  sint32 group_manage = 7;
  sint32 group_admin  = 8;
  sint32 other_use    = 9;
  sint32 other_manage = 10;
  sint32 other_admin  = 11;
  bool recursive      = 12;
}

message ChownRequest
{
  string session_id = 1;
  int32 oid         = 2;
  sint32 user_id    = 3;
  sint32 group_id   = 4;
}

message LockRequest
{
  string session_id = 1;
  int32 oid         = 2;
  int32 level       = 3;
  bool test         = 4;
}

message UnlockRequest
{
  string session_id = 1;
  int32 oid         = 2;
}

message CloneRequest
{
  string session_id = 1;
  int32 oid         = 2;
  string name       = 3;
  bool recursive    = 4;
}

message InstantiateRequest
{
  string session_id     = 1;
  int32 oid             = 2;
  string name           = 3;
  bool hold             = 4;
  string extra_template = 5;
  bool persistent       = 6;
}

message PoolInfoRequest
{
  string session_id  = 1;
  sint32 filter_flag = 2;
  sint32 start       = 3;
  sint32 end         = 4;
}

service TemplateService
{
  rpc Allocate (one.tmpl.AllocateRequest) returns (one.ResponseID);

  rpc Delete (one.tmpl.DeleteRequest) returns (one.ResponseID);

  rpc Info (one.tmpl.InfoRequest) returns (one.ResponseXML);

  rpc Update (one.tmpl.UpdateRequest) returns (one.ResponseID);

  rpc Rename (one.tmpl.RenameRequest) returns (one.ResponseID);

  rpc Chmod (one.tmpl.ChmodRequest) returns (one.ResponseID);

  rpc Chown (one.tmpl.ChownRequest) returns (one.ResponseID);

  rpc Lock (one.tmpl.LockRequest) returns (one.ResponseID);

  rpc Unlock (one.tmpl.UnlockRequest) returns (one.ResponseID);

  rpc Clone (one.tmpl.CloneRequest) returns (one.ResponseID);

  rpc Instantiate (one.tmpl.InstantiateRequest) returns (one.ResponseID);

  rpc PoolInfo (one.tmpl.PoolInfoRequest) returns (one.ResponseXML);
}
